Add optimistic locking support to the memory backend

rosstimothy opened the pull request:
Implements ConditionalUpdate and ConditionalDelete for the memory backend and extends existing backend methods to take the revision into account. The backend test suite has also been updated to run optimistic locking tests on any backends that support it.

Contributes to #30416

espadolini commented on lib/backend/backend.go:
```diff
@@ -422,3 +431,22 @@ func ExactKey(parts ...string) []byte {
+// NewLease creates a lease for the provided [Item].
+func NewLease(item Item) *Lease {
+	if item.Expires.IsZero() {
+		return &Lease{}
```
Don't we want to return the revision every time?

espadolini commented on lib/backend/backend.go:
```diff
@@ -152,20 +157,22 @@ type Batch interface {
+	// Key is the resource identifier.
 	Key []byte
-	// ID is a lease ID, could be empty
+	// ID is a lease ID, could be empty.
 	ID int64
```
Can we mark this ID as deprecated too?

espadolini commented on lib/backend/memory/memory.go:
```diff
@@ -261,6 +264,7 @@ func (m *Memory) PutRange(ctx context.Context, items []backend.Item) error {
 		}
 		if !m.Mirror {
 			event.Item.ID = m.generateID()
+			event.Item.Revision = backend.CreateRevision()
```
Do we want a single revision for the whole `PutRange`? Should we just get rid of `PutRange` altogether? It's only used in the `PutRange` test of the backend suite.

espadolini commented on lib/backend/backend.go:
```diff
@@ -38,6 +39,10 @@ const (
 
+// ErrIncorrectRevision is returned from conditional operations when revisions
+// do not match the expected value.
+var ErrIncorrectRevision = &trace.CompareFailedError{Message: "resource revision does not match, it may have been concurrently created|modified|deleted; please work from the latest state, or use --force to overwrite"}
```
Are we expecting callers to `errors.Is(err, backend.ErrIncorrectRevision)`, as opposed to `trace.IsCompareFailed`? And would that even work just for this specific instance of the error?

espadolini commented on lib/backend/memory/memory.go:
```diff
@@ -401,7 +409,102 @@ func (m *Memory) CompareAndSwap(ctx context.Context, expected backend.Item, repl
+	return backend.NewLease(i), nil
+}
+
+func (m *Memory) ConditionalPut(ctx context.Context, i backend.Item) (*backend.Lease, error) {
```
What's the contract for `ConditionalPut`?

espadolini commented on lib/backend/memory/memory.go:
```diff
@@ -401,7 +409,102 @@ func (m *Memory) CompareAndSwap(ctx context.Context, expected backend.Item, repl
+		return nil, trace.BadParameter("missing parameter key")
+	}
+
+	if m.Mirror {
```
Why would a backend in mirror mode ever get a `ConditionalPut` call?

espadolini commented on lib/backend/backend.go:
```diff
@@ -217,10 +220,13 @@ type Item struct {
 	ID int64
 	// LeaseID is a lease ID, could be set on objects
 	// with TTL
 	LeaseID int64
```
Should we also deprecate `Item.LeaseID`?

espadolini commented on lib/backend/memory/memory.go:
```diff
@@ -317,9 +320,6 @@ func (m *Memory) GetRange(ctx context.Context, startKey []byte, endKey []byte, l
 
 // KeepAlive updates TTL on the lease
 func (m *Memory) KeepAlive(ctx context.Context, lease backend.Lease, expires time.Time) error {
-	if lease.IsEmpty() {
-		return trace.BadParameter("lease is empty")
-	}
```
The other operations on the memory backend check that the key is nonempty, we should probably do the same on `lease.Key` here (or remove the checks in the other operations, assuming that the btree can handle a key of length 0).

espadolini commented on lib/backend/memory/memory.go:
```diff
@@ -374,7 +378,69 @@ func (m *Memory) CompareAndSwap(ctx context.Context, expected backend.Item, repl
+
+	item, found := m.tree.Get(&btreeItem{Item: backend.Item{Key: key}})
+	if !found {
+		return trace.NotFound("key %q is not found", string(key))
```
CompareAndSwap always returns a CompareFailed even if the item doesn't exist; do we want ConditionalDelete and ConditionalUpdate to split the two cases?

fspmarshall commented on lib/backend/memory/memory.go:
```diff
@@ -374,7 +382,69 @@ func (m *Memory) CompareAndSwap(ctx context.Context, expected backend.Item, repl
+
+	item, found := m.tree.Get(&btreeItem{Item: i})
+	if !found {
+		return nil, trace.NotFound("key %q is not found", string(i.Key))
```
Returning a mix of `CompareFailed` and `NotFound` complicates error handling IMO.  The default error message specifically mentions concurrent deletion as a possible source of the error, so I'm not sure there's much benefit from a UX perspective.

More importantly, I think there should be a simple check for "failure due to concurrent modification" that works universally.  Otherwise it will be too easy to accidentally introduce subtle bugs into logic that handles updates.

rosstimothy committed:
Add optimistic locking support to the memory backend

Implements ConditionalUpdate and ConditionalDelete for the
memory backend and extends existing backend methods to take
the revision into account. The backend test suite has also been
updated to run optimisitic locking tests on any backends that
support it.

Contributes to #30416
